hail/src/main/scala/is/hail/expr/ir/lowering/LoweringPass.scala

```diff
@@ -55,7 +55,23 @@ case object LowerTableToDistributedArrayPass extends LoweringPass {
   val after: IRState = CompilableIR
   val context: String = "LowerTableToDistributedArray"
 
-  def transform(ctx: ExecuteContext, ir: BaseIR): BaseIR = LowerTableIR.lower(ir.asInstanceOf[IR])
+  def transform(ctx: ExecuteContext, ir: BaseIR): BaseIR = LowerIR.lower(ir.asInstanceOf[IR], lowerTable = true, lowerBM = false)
+}
+
+case object LowerBlockMatrixToDistributedArrayPass extends LoweringPass {
+  val before: IRState = BlockMatrixOnly
+  val after: IRState = CompilableIR
+  val context: String = "LowerBlockMatrixToDistributedArray"
+
+  def transform(ctx: ExecuteContext, ir: BaseIR): BaseIR = LowerIR.lower(ir.asInstanceOf[IR], lowerTable = false, lowerBM = true)
+}
+
+case object LowerToDistributedArrayPass extends LoweringPass {
```

Comment:
Is there a reason the LowerBlockMatrix and LowerTable passes need to be fused into a single LowerToDistributedArray pass?

Reply:
So the `LowerTableToDistributedArray` pass and the `LowerBlockMatrixToDistributedArray` pass both assert that they're compilable after lowering.

This can pose a problem if a given pipeline has both Tables and BlockMatrices.

I think ideally, I'd like to have `LowerTable...` assert that there aren't any tables left in the lowered thing (or maybe that there are only ValueIR and BlockMatrixIR nodes, which is currently equivalent?), and `LowerBlockMatrix...` assert that there aren't any BlockMatrixIRs remaining post-lowering.

I think if I did that, though, I might either need to restructure how LoweringPass works, or insert a dummy LoweringPass that simply takes the output of a `LowerXToDistributedArray` and checks that it's Compilable to preserve the behavior of the current `LowerTableToDistributedArray`.

I'm also not sure that that's necessarily the right thing to do here, because unlike MatrixTable -> Table which is definitely directional, there's not really a clear order to apply these lowerings---`BlockMatrixToTableApply`, for example, likely wants the Tables lowered first so that it doesn't need to piece together what the BlockMatrix was from a CollectDistributedArray transformation. But something that went in the other direction (I'm imagining eventually abstracting away the MatrixTableToBlockMatrix conversion that we currently stage through write in Python) would likely want the lowering to go in the other direction, so maybe it also makes sense to do all of these lowerings together so that those nodes can implement their lowerings in terms of BlockMatrixStage -> TableStage and vice versa. Maybe this is an argument for introducing an explicit DistributedArray concept in the lowering pipeline/IR structure.

Maybe I should actually just remove this for now, though, and error if we try to lower a pipeline that contains both until we can come up with a better solution.

Reply:
I think that's a good argument for why LowerBlockMatrix... and LowerTable... should be a single lowering pass. So then, is there a reason for keeping the separate passes around (at least at the LoweringPass level)?

Reply:
I want to be able to control whether we lower one vs the other separately right now (from python), as we're kind of actively developing and breaking and improving them. I'm checking the flags in jvmLowerAndExecute which gets called from python's Backend.execute, and I'd kind of like to keep the check there so we have fuller (non-flag-based) control in Scala for tests.

In my mind, we'd rip out the single Table/BlockMatrix lowering passes when we ripped out the flags in Python.

Reply:
I thought about this while I was rebasing and I changed it so that this is now a single LoweringPipeline that's currently parameterized on the types of IR that we want to lower. (Other classes that this affects got changed similarly). Now when we want to rip out the HailContext flags for lowering we just need to rip out this parameter, also.